### PR TITLE
[Config] Read engine args from VllmOmniConfig

### PR DESCRIPTION
**Pull request by Acerak01-fy**
This is PR7 for RFC #4021. It adds the structured engine-argument projection from `VllmOmniConfig` while intentionally leaving current runtime inputs on the existing legacy `list[StageConfig]` / OmegaConf shape. The subsequent AR engine startup (PR8) can switch the consumer path as part of the coordinated stage-init cutover.

## Purpose

`build_engine_args_dict_from_omni_stage_config()` accepts one selected `BaseVllmOmniStageConfig` and builds the flat arguments expected by `OmniEngineArgs` or `OmniDiffusionConfig` directly from its typed sub-configs.It projects model, load, cache, scheduler, runtime, connector, parallel, quantization, and diffusion values without reconstructing them through legacy `engine_args`, OmegaConf, or dict/object compatibility access.

Engine inputs that would otherwise be lost during structured construction are retained by their concern-owned configs. This includes resolved model and tokenizer inputs, worker and platform overrides, connector state, diffusion VAE parallelism, `retains_state_across_chunks`, `logits_processors`, and `enable_session_state_manager`. The structured stage does not gain a catch-all legacy `engine_args` field.

Representation-independent finalization is shared by the legacy and typed implementations. This preserves model/tokenizer path precedence, Audex checkpoint and tokenizer resolution, CLI tokenizer fallback, connector-spec injection, diffusion attention normalization, worker resolution, generation cache defaults, sampling-extra state, and model/platform adjustments.

The engine startup/runtime path is not migrated in this PR. The existing implementation is named `build_legacy_engine_args_dict()`, while the stable `build_engine_args_dict()` entry point preserves its existing signature and
delegates to that legacy implementation. Current production consumers continue to call the stable entry point with legacy stage objects. The typed projection is prepared and tested independently until effective strategy state and the remaining stage-init inputs are ready for the coordinated cutover.

This also updates the related unit tests to cover:

- Typed engine arguments for representative AR, generation, and diffusion stages matching the existing legacy projection.
- Every currently registered pipeline retaining all backend-consumed legacy fields in the typed projection.
- Model and tokenizer subdirectory precedence, CLI tokenizer fallback, and Audex checkpoint behavior.
- Async connector injection, worker resolution, generation cache defaults, sampling-extra state, and diffusion attention normalization.
- Platform-sensitive ROCm attention defaults and diffusion parallel settings.
- The stable compatibility API matching the explicitly named legacy implementation without mutating legacy stage config state.
- Current runtime, headless, and diffusion startup consumers remaining on the stable compatibility entry point.

## Test Plan

vLLM Version: 0.25.0

vLLM-Omni Commit: 41540327

### Test Result

Static checks:

```bash
../.venv-vllm-omni-main/bin/ruff check \
  tests/config/test_omni_config.py \
  tests/engine/test_stage_engine_args.py \
  tests/entrypoints/test_serve.py \
  vllm_omni/config/omni_config.py \
  vllm_omni/engine/stage_init_utils.py \
  vllm_omni/engine/stage_runtime.py \
  vllm_omni/entrypoints/cli/serve.py

../.venv-vllm-omni-main/bin/ruff format --check \
  tests/config/test_omni_config.py \
  tests/engine/test_stage_engine_args.py \
  tests/entrypoints/test_serve.py \
  vllm_omni/config/omni_config.py \
  vllm_omni/engine/stage_init_utils.py \
  vllm_omni/engine/stage_runtime.py \
  vllm_omni/entrypoints/cli/serve.py
```

Result:

```text
All checks passed!
7 files already formatted
```

Unit tests:

```bash
../.venv-vllm-omni-main/bin/python -m pytest \
  tests/config/test_omni_config.py \
  tests/engine/test_arg_utils.py \
  tests/engine/test_async_omni_engine_stage_init.py \
  tests/engine/test_single_stage_mode.py \
  tests/engine/test_stage_engine_args.py \
  tests/engine/test_stage_metadata.py \
  tests/entrypoints/test_serve.py -q
```

Result:

```text
266 passed, 2 skipped, 22 warnings in 45.93s
```

Compatibility check against PR #5172 at `4a5487ba`:

```text
Automatic merge completed without conflicts.
312 passed, 2 skipped, 22 warnings in 41.12s
```

### CODE REVIEW

File: vllm_omni/engine/stage_init_utils.py [hunk 1]
```diff
@@ -748,25 +765,112 @@ def stage_runtime_env(stage_id: int, runtime_cfg: Any) -> Generator[None, None,
...
-    if "tensor_parallel_size" in engine_args and engine_args["tensor_parallel_size"] is None:
-        del engine_args["tensor_parallel_size"]
+    """Copy defined typed config fields into backend adapter kwargs."""
+    projected: dict[str, Any] = {}
+    for config_field in fields(config):
+        name = config_field.name
+        if name in exclude:
+            continue
+        value = getattr(config, name)
+        if value is not None:
```
Comment (hsliuustc0106) [inline vllm_omni/engine/stage_init_utils.py:780]: This projects structured defaults that the legacy adapter leaves unset. For ming_flash_omni_image, legacy omits max_num_seqs (so OmniDiffusionConfig uses 1), but this emits 128 and startup rejects MingImagePipeline because it does not support request batching. The same mismatch changes the vLLM defaults for prefix caching, chunked prefill, hybrid-cache management, and GPU utilization. Please preserve the unset/backend state and compare effective parity in both directions.

File: vllm_omni/config/omni_config.py [hunk 2]
```diff
@@ -751,7 +797,7 @@ def _global_stage_cli_fields() -> frozenset[str]:
         frozenset(f.name for f in fields(OmniEngineArgs))
```
Comment (hsliuustc0106) [inline vllm_omni/config/omni_config.py:901]: This accepts inherited OmniEngineArgs options globally, but the later per-concern allowlists silently discard options without an owner. I reproduced revision, tokenizer_revision, code_revision, seed, and download_dir surviving the legacy AR path and disappearing from the typed path. Please make ownership exhaustive or reject unowned fields instead of accepting and dropping them.

File: vllm_omni/config/omni_config.py [hunk 3]
```diff
@@ -1116,12 +1196,27 @@ def _build_quantization_config(
...
+        kwargs["trust_remote_code"] = _copy_value(deploy.trust_remote_code)
+    if "dtype" not in kwargs and deploy.dtype is not None:
+        kwargs["dtype"] = _copy_value(deploy.dtype)
+    if "active_stream_window" not in kwargs:
+        kwargs["active_stream_window"] = _copy_value(deploy.active_stream_window)
+    if "custom_voice_dir" not in kwargs and deploy.custom_voice_dir is not None:
+        kwargs["custom_voice_dir"] = _copy_value(deploy.custom_voice_dir)
     if "has_sampling_extra_args" not in kwargs:
         kwargs["has_sampling_extra_args"] = bool((default_sampling_params or {}).get("extra_args"))
     if "model_subdir" not in kwargs and topology.model_subdir is not None:
```
Comment (hsliuustc0106) [inline vllm_omni/config/omni_config.py:1354]: The legacy builder lets stage engine_extras override topology model_subdir and tokenizer_subdir, but those keys never reach engine here, so this fallback restores the topology values. A direct reproduction selected /tmp/override-model through legacy and /tmp/top-model through typed. Please preserve deploy precedence and cover both subdirectory overrides.

File: vllm_omni/config/omni_config.py [hunk 4]
```diff
@@ -1158,21 +1264,35 @@ def _build_scheduler_config(
...
-def _build_connector_config(stage_deploy: StageDeployConfig | None) -> OmniStageConnectorConfig:
+def _build_connector_config(
+    deploy: DeployConfig,
+    topology: StagePipelineConfig,
+    stage_deploy: StageDeployConfig | None,
+    engine: _ConnectorEngineOverrides,
+) -> OmniStageConnectorConfig:
     output_connectors = stage_deploy.output_connectors if stage_deploy is not None else None
     input_connectors = stage_deploy.input_connectors if stage_deploy is not None else None
+    omni_kv_config = _first_defined(
```
Comment (hsliuustc0106) [inline vllm_omni/config/omni_config.py]: This reverses the legacy omni_kv_config precedence: legacy applies topology after stage extras, while this selects the stage value first. With topology need_recv_cache=True and deploy need_send_cache=True, legacy emits receive and typed emits send. Please keep the legacy ordering; if the precedence change is intentional, document and test it as a behavior change.

Comment (Acerak01-fy) [reply vllm_omni/config/omni_config.py]: Done.

File: vllm_omni/config/omni_config.py (hunk 3 above)
Comment (Acerak01-fy) [reply vllm_omni/config/omni_config.py:1354]: Done.

File: vllm_omni/config/omni_config.py (hunk 2 above)
Comment (Acerak01-fy) [reply vllm_omni/config/omni_config.py:901]: Done.

File: vllm_omni/engine/stage_init_utils.py (hunk 1 above)
Comment (Acerak01-fy) [reply vllm_omni/engine/stage_init_utils.py:780]: Done.

File: vllm_omni/config/omni_config.py [hunk 5]
```diff
@@ -739,21 +814,51 @@ def enrich_config(self) -> None:
...
+    | _PARALLEL_ENGINE_FIELDS
+    | _DIFFUSION_STAGE_ENGINE_FIELDS
+)
+
+
+def _validate_stage_cli_override_ownership(
+    stage_id: int,
+    overrides: Mapping[str, Any],
+) -> None:
+    unowned_fields = set(overrides) - _STAGE_ENGINE_FIELDS
```
Comment (hsliuustc0106) [inline vllm_omni/config/omni_config.py]: This validates only CLI inputs and uses an all-stage union. Deploy `engine_extras` are never checked, and LLM validation accepts diffusion-only fields: deploy `enable_lora=True` becomes `False`, while LLM CLI `kv_cache_dtype=fp8` becomes `auto` through the typed path. Please validate the merged input against execution-type-specific owners before partitioning.

File: tests/engine/test_stage_engine_args.py [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+
+        comparable_fields = (legacy_args.keys() & backend_fields) - {
+            "cache_config",
+            "parallel_config",
+        }
+        assert {name: typed_args[name] for name in comparable_fields} == {
+            name: legacy_args[name] for name in comparable_fields
+        }
+
+        deferred_default_fields = backend_fields & _BACKEND_DEFERRED_DEFAULT_FIELDS
```
Comment (hsliuustc0106) [inline tests/engine/test_stage_engine_args.py]: This only compares the six hard-coded deferred defaults. Adding typed-only `skip_tokenizer_init=True` still lets the parity tests pass. Please compare effective values across all backend fields, or reject unexpected typed-only keys.

**Comment by hsliuustc0106** [comment]
Can we align this PR with [#5140’s final target](https://github.com/vllm-project/vllm-omni/pull/5140#final-target-layered-configuration-structure), or document why this PR supersedes it?

#5140 gives each resolved stage Omni-owned topology/runtime/connector/request metadata plus one `engine_spec`: `EngineArgs` for AR/generation or an un-enriched `OmniDiffusionConfig` for diffusion, materialized in the owning process.

This PR instead splits selected `EngineArgs` fields across manual `OmniStage*Config` trees and `_..._FIELDS` sets, then flattens them back. Those buckets already differ from upstream ownership—`max_model_len`, tokenizer fields, and hybrid-cache settings are examples—and the inline threads show the result: accepted fields can be lost, execution-type validation can use the wrong owner set, and parity tests can miss typed-only defaults.

Since runtime has not cut over, please either rework the projector around one complete engine spec, or make the divergence explicit in #4021 and add a schema-derived exhaustive support check against the pinned upstream `EngineArgs`.

File: tests/engine/test_stage_engine_args.py (hunk 6 above)
Comment (Acerak01-fy) [reply tests/engine/test_stage_engine_args.py]: Done

**Review by hsliuustc0106 — APPROVED** [review]
lgtm
